File: cpp/LeetCode2901.cpp
class Solution {
public:
    vector<string> getWordsInLongestSubsequence(vector<string>& words, vector<int>& groups) {
        int n = size(words);
        vector<int> dp(n, 1);
        vector<int> prev(n, -1);
        int max_index = 0;

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (checkHamming(words[i], words[j]) && groups[i] != groups[j]) {
                    if (dp[j] + 1 > dp[i]) {
                        dp[i] = dp[j] + 1;
                        prev[i] = j;
                    }
                    if (dp[i] > dp[max_index]){
                        max_index = i;
                    }
                }
            }
        }

        vector<string> ans;
        for (int i = max_index; i > -1; i = prev[i]) {
            ans.emplace_back(words[i]);
        }
        reverse(ans.begin(), ans.end());

        return ans;
    }

    bool checkHamming(string v, string w) {
        return hammingDistance(v, w) == 1;
    }

    int hammingDistance(string v, string w) {
        if (v.size() != w.size()) {
            return -1;
        } 
        int count = 0;
        for (int i = 0; i < v.size(); i++) {
            if (v[i] != w[i]) {
                count++;
            }
        }
        return count;
    }
};

File: python/LeetCode2901.py
class Solution:
    def getWordsInLongestSubsequence(
        self, words: List[str], groups: List[int]
    ) -> List[str]:
        n = len(words)
        dp = [1] * n
        prev = [-1] * n
        max_index = 0

        for i in range(1, n):
            for j in range(i):
                if self.checkHamming(words[i], words[j]) and groups[i] != groups[j]:
                    if dp[j] + 1 > dp[i]:
                        dp[i] = dp[j] + 1
                        prev[i] = j
                    if dp[i] > dp[max_index]:
                        max_index = i

        result = []
        i = max_index
        while i > -1:
            result.append(words[i])
            i = prev[i]
        result.reverse()
        return result

    def checkHamming(self, v: str, w: str) -> bool:
        return self.hammingDistance(v, w) == 1

    def hammingDistance(self, v: str, w: str) -> int:
        if len(v) != len(w):
            return -1
        distance = 0
        for c, s in zip(v, w):
            if c != s:
                distance += 1
        return distance
